### PR TITLE
executor: Use multiple goroutines to do join

### PR DESCRIPTION
**Pull request by shenli**

### CODE REVIEW

**Comment by hanfei1991** [comment]
@tiancaiamao @zimulala @coocood

File: expression/expression.go [hunk 1]
```diff
@@ -303,14 +331,32 @@ func (sf *ScalarFunction) GetType() *types.FieldType {
...
+		arg.InitArgsBuffer(num)
+	}
+}
+
+// EvalInShard implements Expression interface.
+func (sf *ScalarFunction) EvalInShard(row []types.Datum, ctx context.Context, shardID int) (types.Datum, error) {
+	args := sf.argsBuffer[shardID]
 	for i, arg := range sf.Args {
-		sf.ArgValues[i], err = arg.Eval(row, ctx)
+		//sf.ArgValues[i], err = arg.Eval(row, ctx)
```
Comment (shenli) [inline expression/expression.go]: Useless code.

File: executor/new_executor.go [hunk 2]
```diff
@@ -149,12 +223,86 @@ func (e *HashJoinExec) prepare() error {
...
+}
+
+func (e *HashJoinExec) join(idx int, bigRow *Row) bool {
+	var (
+		matchedRows []*Row
+		err         error
+	)
+	bigMatched := true
+	if e.bigFilter != nil {
+		bigMatched, err = expression.EvalBoolInShard(e.bigFilter, bigRow.Data, e.ctx, idx)
```
Comment (shenli) [inline executor/new_executor.go]: There is another way to keep concurrent safe.
1. Create a struct named execContext which contains: e.bigTableRows[idx], deep cloned e.bigFilter, e.hashKeyBuffer[idx]
2. Each doJoin goroutine get an execContext instance as parameter.

File: executor/new_executor.go [hunk 3]
```diff
@@ -45,17 +46,34 @@ type HashJoinExec struct {
...
 	targetTypes []*types.FieldType
+
+	finished bool
+	wg       sync.WaitGroup // for sync multiple join workers
+
+	// Buffer
+	datumBuffer   [][]types.Datum
+	hashKeyBuffer [][]byte
+
+	// Concurrent exec
```
Comment (zimulala) [inline executor/new_executor.go]: Unify the first letter format.

File: executor/new_executor.go [hunk 4]
```diff
@@ -149,12 +223,86 @@ func (e *HashJoinExec) prepare() error {
...
 }
 
-func (e *HashJoinExec) constructMatchedRows(bigRow *Row) (matchedRows []*Row, err error) {
-	hasNull, hashcode, err := getHashKey(e.bigHashKey, bigRow, e.targetTypes)
+func (e *HashJoinExec) closeChanWorker() {
+	e.wg.Wait()
+	close(e.resultRows)
+}
+
+// do join job
```
Comment (zimulala) [inline executor/new_executor.go]: Finish the comment.

File: executor/new_executor.go [hunk 5]
```diff
@@ -45,17 +46,34 @@ type HashJoinExec struct {
...
+	finished bool
+	wg       sync.WaitGroup // for sync multiple join workers
+
+	// Buffer
+	datumBuffer   [][]types.Datum
+	hashKeyBuffer [][]byte
+
+	// Concurrent exec
+	concurrency  int
+	bigTableRows []chan []*Row // channle for join workers to get big table rows
```
Comment (tiancaiamao) [inline executor/new_executor.go]: s/channle/channel

**Comment by tiancaiamao** [comment]
`make race` would fail...
this PR is not a simple change, we must make sure it won't bring concurrency issues

File: executor/new_executor.go [hunk 6]
```diff
@@ -112,7 +129,64 @@ func (e *HashJoinExec) Fields() []*ast.ResultField {
...
 func (e *HashJoinExec) prepare() error {
+	e.finished = false
+	e.concurrency = 5
+	e.datumBuffer = make([][]types.Datum, e.concurrency)
+	e.hashKeyBuffer = make([][]byte, e.concurrency)
+	for i := range e.datumBuffer {
+		e.datumBuffer[i] = make([]types.Datum, len(e.bigHashKey))
+	}
+	for i := range e.hashKeyBuffer {
+		e.hashKeyBuffer[i] = make([]byte, 0, 10000)
```
Comment (zimulala) [inline executor/new_executor.go]: To define 100, 1000, 10000 as constant make them clearer.

**Comment by hanfei1991** [comment]
Some data races will never happen. Like reading map and writing map, this will be granteed by logic. @tiancaiamao

**Comment by ngaut** [comment]
We should pass the race detector. Kill any warnings whether they happened or not.

**Comment by tiancaiamao** [comment]
if data races never happen, `go test -race` should not detect it.
once concurrency issues are introduced, fix them would take much more effort.
@hanfei1991

File: executor/new_executor.go [hunk 7]
```diff
@@ -149,12 +217,86 @@ func (e *HashJoinExec) prepare() error {
...
+func (e *HashJoinExec) doJoin(idx int) {
+	for {
+		var (
+			bigRows []*Row
+			ok      bool
+			err     error
+		)
+		select {
+		case bigRows, ok = <-e.bigTableRows[idx]:
+		case err = <-e.bigTableErr:
```
Comment (zimulala) [inline executor/new_executor.go]: If find a bigTableErr it only returns one goroutine, other 4 goroutines don't return.

Comment (shenli) [inline executor/new_executor.go]: See line 365.
If any error found in  resultErr, e.finished will be set to true.

Comment (hanfei1991) [inline executor/new_executor.go]: She means the finished flag only influence the big fetch executor, but won't let the do join threads stop working until they get out all the data in channel.

Comment (shenli) [inline executor/new_executor.go]: See line 256.

Comment (hanfei1991) [inline executor/new_executor.go]: I added e.finished to condition in line 256 in commit 5e8e764 ....

**Comment by shenli** [comment]
@hanfei1991 `make race` passed?

**Comment by hanfei1991** [comment]
@shenli not yet

File: executor/new_executor.go [hunk 8]
```diff
@@ -112,7 +133,55 @@ func (e *HashJoinExec) Fields() []*ast.ResultField {
...
+			close(cn)
+		}
+	}()
+	for {
+		if e.finished {
+			break
+		}
+		rows := make([]*Row, 0, batchSize)
+		done := false
+		for i := 0; i < batchSize; i++ {
```
Comment (shenli) [inline executor/new_executor.go]: It is better to use a slow start algorithm for batchSize. For example the batchSize is 1, 2, 4, ...., maxBatchSize for each loop. The join worker will get data earlier. @hanfei1991

File: executor/new_executor.go [hunk 9]
```diff
@@ -45,22 +46,38 @@ type HashJoinExec struct {
...
+	bigTableRows     []chan []*Row
+	bigTableErr      chan error
+	hashJoinContexts []*hashJoinCtx
+
+	// Channels for output.
+	resultErr  chan error
+	resultRows chan *Row
+}
+
+type hashJoinCtx struct {
```
Comment (shenli) [inline executor/new_executor.go]: Add  a few comments for this structure.

**Comment by shenli** [comment]
LGTM

**Pull request merged**